File: StkExtensionPlugins/PluginScripts/Perl/TransmitterModel.pl
##########################################################################################
# SAMPLE FOR PERL BASED CUSTOM TRANSMITTER MODEL PLUGIN SCRIPT PROVIDED BY THE USER
# PLEASE ADD YOUR MODEL IN THE USER TRANSMITTER MODEL AREA BELOW.
# DO NOT CHANGE ANYTHING ELSE IN THE SCRIPT
##########################################################################################

# Perl_TransmitterModel

# declare some global variablesfunction

my $Perl_TransmitterModel_compute_init = -1;
my $Perl_TransmitterModel_Inputs, $Perl_TransmitterModel_Outputs;
my %Perl_TransmitterModel_Outputs_ArgHash;

sub Perl_TransmitterModel
{
# the inputs to the the script arise as a reference to an array
	# the STKUtil::getInputArray function is used to get at the array itself

	my @inputData = @{$_[0]};
 
	my @retVal;

	if ( !defined($inputData[0]) )
	{
		# do compute

		@retVal = Perl_TransmitterModel_compute(@inputData);

	}
	elsif ( $inputData[0] eq 'register' )
	{
		$Perl_TransmitterModel_compute_init = -1;

		@retVal = Perl_TransmitterModel_register();

      }
	elsif ( $inputData[0] eq 'compute' )
	{
		@retVal = Perl_TransmitterModel_compute(@inputData);
      }
	else
	{
		# error: do nothing
	}

	# MUST return a reference to an array, as shown below

	return \@retVal;
}

sub Perl_TransmitterModel_register
{
	my @argStr;

	push @argStr, "ArgumentType = Output; Name = Frequency;         ArgumentName = Frequency";
	push @argStr, "ArgumentType = Output; Name = Power;             ArgumentName = Power";
	push @argStr, "ArgumentType = Output; Name = Gain;              ArgumentName = Gain";
	push @argStr, "ArgumentType = Output; Name = DataRate;          ArgumentName = DataRate";
	push @argStr, "ArgumentType = Output; Name = Bandwidth;         ArgumentName = Bandwidth";
	push @argStr, "ArgumentType = Output; Name = Modulation;        ArgumentName = Modulation";
	push @argStr, "ArgumentType = Output; Name = PostTransmitLoss;  ArgumentName = PostTransmitLoss";
	push @argStr, "ArgumentType = Output; Name = PolType;           ArgumentName = PolType";
	push @argStr, "ArgumentType = Output; Name = PolRefAxis;        ArgumentName = PolRefAxis";
	push @argStr, "ArgumentType = Output; Name = PolTiltAngle;      ArgumentName = PolTiltAngle";
	push @argStr, "ArgumentType = Output; Name = PolAxialRatio;     ArgumentName = PolAxialRatio";
	push @argStr, "ArgumentType = Output; Name = UseCDMASpreadGain; ArgumentName = UseCDMASpreadGain";
	push @argStr, "ArgumentType = Output; Name = CDMAGain;          ArgumentName = CDMAGain";

	push @argStr, "ArgumentType = Input;  Name = EpochSec;          ArgumentName = EpochSec";
	push @argStr, "ArgumentType = Input;  Name = DateUTC;           ArgumentName = DateUTC";
	push @argStr, "ArgumentType = Input;  Name = CbName;            ArgumentName = CbName";
	push @argStr, "ArgumentType = Input;  Name = XmtrPosCBF;        ArgumentName = XmtrPosCBF";
	push @argStr, "ArgumentType = Input;  Name = XmtrAttitude;      ArgumentName = XmtrAttitude";
	push @argStr, "ArgumentType = Input;  Name = RcvrPosCBF;        ArgumentName = RcvrPosCBF";
	push @argStr, "ArgumentType = Input;  Name = RcvrAttitude;      ArgumentName = RcvrAttitude";
	
	return @argStr; 
}


sub Perl_TransmitterModel_compute
{
	# the inputs here are in the order of the requested Inputs, as registered
	my @origArray = @_;

	# $origArray[0] is the calling mode

	# next argument is Time and is a value. Values are simply scalars in Perl so the assignment is easy:	
	my $time = $origArray[1];

	# next argument is Date and is a String. Strings are simply scalars in Perl so the assignment is easy:	
	my $date = $origArray[2];

	# next argument is CbName and is a String. 	
	my $CbName = $origArray[3];
	
	# next argument is named XmtrPosCBF and is type (CartVec3) Array and is double-3. 
		
	my @XmtrPosCBF = @{$origArray[4]};
		
	# next argument is named XmtrAttitude and is type (Quaternion) Array and is double-4. 
		
	my @XmtrAttitude = @{$origArray[5]};
	
	# next argument is named RcvrPosCBF and is type (CartVec3) Array and is double-3.
	
        my @RcvrPosCBF = @{$origArray[6]};
	
	# next argument is named RcvrAttitude and is type (Quaternion) Array and is double-4. 
	
	my @RcvrAttitude = @{$origArray[7]};
	
	if($Perl_TransmitterModel_compute_init < 0)
	{
		$Perl_TransmitterModel_compute_init = 1; 

		# The following hashes have been created automatically after this script has registered its inputs and outputs.
		# Each hash contains information about the arguments for this script. The hashes have been created as a
		# user convenience, for those users wanting to know, during the running of the script, what the inputs
		# and outputs are. In many cases, the script write doesn't care, in which case this entire if-block
		# is unneeded and can be removed.

		$Perl_TransmitterModel_Inputs = $g_PluginArrayInterfaceHash{'Perl_TransmitterModel'}{'Inputs'};
		$Perl_TransmitterModel_Outputs = $g_PluginArrayInterfaceHash{'Perl_TransmitterModel'}{'Outputs'};

		%Perl_TransmitterModel_Outputs_ArgHash = $Perl_TransmitterModel_Outputs->getArgumentHash();

		# comment out the line below if you don't want to see the inputs and outputs each time the script is run
		Perl_TransmitterModel_showArgs();
	}

	# NOTE: the outputs that are returned MUST be in the same order as registered
	
	#############################################################################################
	# USER TRANSMITTER MODEL AREA.
	# PLEASE REPLACE THE CODE BELOW WITH YOUR TRANSMITTER MODEL
	#############################################################################################

	# compute the Test Model : Transmitter Model
	

	$fromX = @XmtrPosCBF[0];
	$fromY = @XmtrPosCBF[1];
	$fromZ = @XmtrPosCBF[2];
	
	$toX = @RcvrPosCBF[0];
	$toY = @RcvrPosCBF[1];
	$toZ = @RcvrPosCBF[2];

	$rX = $fromX - $toX;
	$rY = $fromY - $toY;
	$rZ = $fromZ - $toZ;
	
	$xmA = @XmtrAttitude[0];
	$xmB = @XmtrAttitude[1];
	$xmC = @XmtrAttitude[2];
	$xmD = @XmtrAttitude[3];
	
	$rcA = @RcvrAttitude[0];
	$rcB = @RcvrAttitude[1];
	$rcC = @RcvrAttitude[2];
	$rcD = @RcvrAttitude[3];
	
	$range = sqrt ($rX * $rX + $rY * $rY + $rZ * $rZ);

	# this defines the return array
	my @returnArray = ();

	# RETURN YOUR RESULTS BELOW

	$returnArray[0] =  12.3e9;
	$returnArray[1] =  35;
	$returnArray[2] =  30;
	$returnArray[3] =  12.0e6;
	$returnArray[4] =  25.0e6;
	$returnArray[5] =  "BPSK";
	$returnArray[6] =  -2.3;
	$returnArray[7] =  0;
	$returnArray[8] =  0;
	$returnArray[9] =  0.0;
	$returnArray[10] = 1.0;
	$returnArray[11] = 0;
	$returnArray[12] = 0.0;
	
	#############################################################################################
	# END OF USER MODEL AREA	
	#############################################################################################

	return @returnArray;
}


sub Perl_TransmitterModel_showArgs
{
	my @argStrArray;

	STKUtil::printOut "Doing Perl_TransmitterModel_compute_init\n";

	@argStrArray = ();

	push @argStrArray, $Perl_TransmitterModel_Inputs->{'FunctionName'}->{'Name'} . " Inputs \n";

	# the first arg on input is the calling mode

	push @argStrArray, "0 : this is the calling mode\n";

	my @args = $Perl_TransmitterModel_Inputs->getArgumentArray();

	# to see description args

	my $index, $descrip;

	foreach $arg (@args)
	{
		($index, $descrip) = $Perl_TransmitterModel_Inputs->getArgument($arg);

		push @argStrArray, "$index : $arg = $descrip\n";
	}

	STKUtil::printOut @argStrArray;

	@argStrArray = ();

	push @argStrArray, $Perl_TransmitterModel_Outputs->{'FunctionName'}->{'Name'} . " Outputs \n";

	my @args = $Perl_TransmitterModel_Outputs->getArgumentArray();

	# to see description args

	my $index, $descrip;

	foreach $arg (@args)
	{
		($index, $descrip) = $Perl_TransmitterModel_Outputs->getArgument($arg);

		push @argStrArray, "$index : $arg = $descrip\n";
	}

	STKUtil::printOut @argStrArray;

}

# MUST end Perl script file with a non-zero integer

1;

File: StkExtensionPlugins/PluginScripts/Matlab/TransmitterModel.m
function [output] = Matlab_TransmitterModel(input)
	% NOTE: the outputs that are returned
	%       MUST be in the same order as registered

	% NOTE: Please DO NOT change anything above or below the USER MODEL AREA.
	% 
	% This example shows static output parameters, but these parameters
      % may change value at each time step.

switch input.method
    
    case 'register'
 	% register output variables       
	
	% The frequency of the transmitter carrier is a double in Hz.
        Frequency= {'ArgumentName','Frequency',...
                     'Name','Frequency',...
                     'ArgumentType','Output'};

	% The final output power is a double in dBW.
        Power= {'ArgumentName','Power',...
                     'Name','Power',...
                     'ArgumentType','Output'};
					 
	% The gain of the radiating elements of the antenna is a double in dB or dBi.
        Gain= {'ArgumentName','Gain',...
                     'Name','Gain',...
                     'ArgumentType','Output'};
					 
	% The information bit rate is a double in bits per second.
        DataRate= {'ArgumentName','DataRate',...
                    'Name','DataRate',...
                    'ArgumentType','Output'};
					
	% The bandwidth of the RF spectrum is a double in Hz.
       Bandwidth= {'ArgumentName','Bandwidth',...
                    'Name','Bandwidth',...
                    'ArgumentType','Output'};
 
	% The type of modulation used by the transmitter is a String. This must be one of the STK registered modulation types. However,
	% users can add their own modulation types, and STK will register them (see online Help for the Comm module).
	   Modulation= {'ArgumentName','Modulation',...
                     'Name','Modulation',...
                     'ArgumentType','Output'};
					 
	% A collection of post-transmit losses, e.g. antenna dish coupling or radome loss is a double in dB.
        PostTransmitLoss= {'ArgumentName','PostTransmitLoss',...
                    'Name','PostTransmitLoss',...
                    'ArgumentType','Output'};
					
	% The type of polarization is an Integer:
	% Value 	Polarization Type 		Required Parameter(s)
	%	0 		None 					None
	%	1 		Linear 					Reference Axis
	%	2		Right Hand Circular 	None
	% 	3		Left Hand Circular 		None
	%	4 		Vertical 				Reference Axis, Tilt Angle
	%	5 		Horizontal 				Reference Axis, Tilt Angle
	%	6 		Elliptical 				Reference Axis, Tilt Angle, Axial Ratio
    
        PolType= {'ArgumentName','PolType',...
                    'Name','PolType',...
                    'ArgumentType','Output'};
	
	%	The polarization reference axis is an Integer that is used to align transmitter polarization to receiver polarization,	
	%	with 0, 1 and 2 representing the X, Y and Z axes, respectively.
        PolRefAxis= {'ArgumentName','PolRefAxis',...
                     'Name','PolRefAxis',...
                     'ArgumentType','Output'};
					 
    
	%  The polarization tilt angle measured from the reference axis is a double in degrees.
        PolTiltAngle= {'ArgumentName','PolTiltAngle',...
                    'Name','PolTiltAngle',...
                    'ArgumentType','Output'};

	%  The polarization axial ratio is a real.
        PolAxialRatio= {'ArgumentName','PolAxialRatio',...
                     'Name','PolAxialRatio',...
                     'ArgumentType','Output'};
    
    %  UseCDMASpreadGain is a flag (0 or 1) that specifies whether or not to use the bandwidth spreading gain is a Boolean.
		UseCDMASpreadGain= {'ArgumentName','UseCDMASpreadGain',...
                    'Name','UseCDMASpreadGain',...
                    'ArgumentType','Output'};
    
	%	The CDMA coding gain advantage is a double in dB.
        CDMAGain= {'ArgumentName','CDMAGain',...
                    'Name','CDMAGain',...
                    'ArgumentType','Output'};

 	% register input variables       
   
    % The current date and time is a String variable. 
        date = {'ArgumentName','DateUTC',...
                'Name','DateUTC',...
                'ArgumentType','Input'};
 
    % The scenario central body is a String variable. 
	   CbName = {'ArgumentName','CbName',...
                  'Name','CbName',...
                  'ArgumentType','Input'};
				  
	% The scenario simulation epoch time is a double in seconds.
	   EpochSec = {'ArgumentName','EpochSec',...
                'Name','EpochSec',...
                'ArgumentType','Input'};
				  
	% The transmitter position in Central Body Fixed coordinates is a vector of doubles, of length 3, corresponding to the X, Y 
	% and Z values with the unit of meters.
        XmtrPosCBF = {'ArgumentName','XmtrPosCBF',...
               'Name','XmtrPosCBF',...
               'ArgumentType','Input'};

	% The attitude quaternion of the transmitter is a vector of doubles, of length 4.
	   XmtrAttitude= {'ArgumentName','XmtrAttitude',...
               'Name','XmtrAttitude',...
               'ArgumentType','Input'};

	% The receiver position in Central Body Fixed coordinates is a vector of doubles, of length 3, corresponding to the X, Y and Z
	% values with the unit of meters.       
	   RcvrPosCBF= {'ArgumentName','RcvrPosCBF',...
               'Name','RcvrPosCBF',...
               'ArgumentType','Input'};
 
	% The attitude quaternion of the receiver is a vector of doubles, of length 4.
        RcvrAttitude= {'ArgumentName','RcvrAttitude',...
               'Name','RcvrAttitude',...
               'ArgumentType','Input'};
   
        output = {Frequency, Power, Gain, DataRate, Bandwidth,...
                  Modulation, PostTransmitLoss,...
                  PolType, PolRefAxis, PolTiltAngle, PolAxialRatio,...
                  UseCDMASpreadGain, CDMAGain,...
                  date, CbName, EpochSec, XmtrPosCBF, XmtrAttitude, RcvrPosCBF, RcvrAttitude};
    
    case 'compute'
    
        computeData = input.methodData;

	% compute the Test Model : 
	% Example Model for testing only
	% Transmitter input & outpur parameter usage is shown
	% 

	% USER Transmitter MODEL AREA.
        
	  xmtrPos = [ 0 0 0 ];
	  rcvrPos = [ 0 0 0 ];
        xmAttQuat = [ 0 0 0 0 ];
        rcAttQuat = [ 0 0 0 0 ];
	  xmtrPos = computeData.XmtrPosCBF;
	  rcvrPos = computeData.RcvrPosCBF;
        xmAttQuat = computeData.XmtrAttitude;
        rcAttQuat = computeData.RcvrAttitude;

	  range = norm(xmtrPos  - rcvrPos);

        output.Frequency= 12.3e9;
        output.Power= 35;
        output.Gain= 30;
        output.DataRate= 12.0e6;
        output.Bandwidth= 25.0e6;
        output.Modulation= 'BPSK';
        output.PostTransmitLoss= -2.3;
        output.PolType= 0;
        output.PolRefAxis= 0;
        output.PolTiltAngle= 0.0;
        output.PolAxialRatio= 1.0;
        output.UseCDMASpreadGain= 0;
        output.CDMAGain= 0.0;

	% END OF USER MODEL AREA

    otherwise
        output = [];
end

